src/passes/Heap2Local.cpp

```diff
@@ -430,6 +430,16 @@ struct Heap2LocalOptimizer {
       replaceCurrent(curr->value);
     }
 
+    void visitRefCast(RefCast* curr) {
+      if (!reached.count(curr)) {
+        return;
+      }
+
+      // It is safe to optimize out this RefCast, since we proved it
+      // contains our allocation, and so cannot trap.
```

Comment:
This comment is identical to the above for RefAs, but I think we need more here. That the RefAsNonNull receives an allocation is enough to prove it is not null, but here we need the extra type checking on line 663. So perhaps add `, and we verified already that it receives the correct type`.